Plugins/ConnectPluginUtilities/Generator.swift

```diff
@@ -58,30 +79,75 @@ open class Generator {
             self.printLine(comments)
         }
     }
 
-    public func printFilePreamble() {
-        self.printLine("// Code generated by protoc-gen-connect-swift. DO NOT EDIT.")
-        self.printLine("//")
-        self.printLine("// Source: \(self.descriptor.name)")
-        self.printLine("//")
-        self.printLine()
-    }
-
     public func printModuleImports(adding additional: [String] = []) {
         let defaults = ["Connect", "Foundation", self.options.swiftProtobufModuleName]
         let extraOptionImports = self.options.extraModuleImports
-        let mappings = self.options.protoToModuleMappings
-            .neededModules(forFile: self.descriptor) ?? []
-        let allImports = (defaults + mappings + extraOptionImports + additional).sorted()
-
+        let allImports = (defaults + self.neededModules + extraOptionImports + additional).sorted()
         for module in allImports {
             self.printLine("import \(module)")
         }
     }
 }
+
+extension Generator: SwiftProtobufPluginLibrary.CodeGenerator {
+    private func resetAndPrintFile(
+        for descriptor: SwiftProtobufPluginLibrary.FileDescriptor
+    ) -> String {
+        self.namer = SwiftProtobufPluginLibrary.SwiftProtobufNamer(
+            currentFile: descriptor,
+            protoFileToModuleMappings: self.options.protoToModuleMappings
+        )
+        self.neededModules = self.options.protoToModuleMappings
+            .neededModules(forFile: descriptor) ?? []
+        self.services = descriptor.services
+        self.printer = SwiftProtobufPluginLibrary.CodePrinter(indent: "    ".unicodeScalars)
+        self.printContent(for: descriptor)
+        return self.printer.content
+    }
+
+    public func generate(
+        files: [SwiftProtobufPluginLibrary.FileDescriptor],
+        parameter: any SwiftProtobufPluginLibrary.CodeGeneratorParameter,
+        protoCompilerContext _: any SwiftProtobufPluginLibrary.ProtoCompilerContext,
+        generatorOutputs: any SwiftProtobufPluginLibrary.GeneratorOutputs
+    ) throws {
+        self.options = try GeneratorOptions(commandLineParameters: parameter)
+        guard self.options.generateAsyncMethods || self.options.generateCallbackMethods else {
+            throw GeneratorError(
+                message: "Either async methods or callback methods must be enabled"
+            )
+        }
+
+        for descriptor in files where !descriptor.services.isEmpty {
+            try generatorOutputs.add(
+                fileName: FilePathComponents(path: descriptor.name).outputFilePath(
+                    withExtension: self.outputFileExtension,
+                    using: self.options.fileNaming
+                ),
+                contents: self.resetAndPrintFile(for: descriptor)
+            )
+        }
+    }
+
+    public var supportedEditionRange: ClosedRange<SwiftProtobuf.Google_Protobuf_Edition> {
+        return SwiftProtobufPluginLibrary.DescriptorSet.bundledEditionsSupport
```

Comment:
This seems potentially dangerous: a future edition _could_ introduce service or method-specific changes that need to be handled by the plugin. In such a case, if this plugin code (without the necessary updates) were paired with a newer version of the runtime (that _was_ updated to support that edition) then it would report that the edition is supported even though it may generate incorrect code.

While it's a bit of a maintenance nuisance to have to manually bump the max supported edition, I think that's the safest way to implement this.

Reply:
Fair point, updated